core/mempool/src/pool_task.rs

```diff
@@ -0,0 +1,167 @@
+use std::sync::Arc;
+use std::time::Duration;
+
+use futures::future;
+use futures::sink::Sink;
+use futures::stream::Stream;
+use futures::sync::mpsc::{Receiver, Sender};
+use futures::Future;
+use log::{error, info, warn};
+use tokio::{self, timer::Interval};
+
+use primitives::aggregate_signature::{BlsPublicKey, BlsSecretKey};
+use primitives::chain::{ChainPayload, PayloadRequest, PayloadResponse};
+use primitives::hash::CryptoHash;
+use primitives::signature::{PublicKey, SecretKey};
+use primitives::types::AuthorityId;
+
+use nightshade::nightshade_task::Control;
+
+use crate::Pool;
+
+#[derive(Clone, Debug)]
+pub enum MemPoolControl {
+    Reset {
+        authority_id: AuthorityId,
+        num_authorities: usize,
+
+        owner_uid: u64,
+        block_index: u64,
+        public_keys: Vec<PublicKey>,
+        owner_secret_key: SecretKey,
+        bls_public_keys: Vec<BlsPublicKey>,
+        bls_owner_secret_key: BlsSecretKey,
+    },
+    Stop,
+}
+
+pub fn spawn_pool(
+    pool: Arc<Pool>,
+    mempool_control_rx: Receiver<MemPoolControl>,
+    control_tx: Sender<Control>,
+    retrieve_payload_rx: Receiver<(AuthorityId, CryptoHash)>,
+    payload_announce_tx: Sender<(AuthorityId, ChainPayload)>,
+    payload_request_tx: Sender<PayloadRequest>,
+    payload_response_rx: Receiver<PayloadResponse>,
+    payload_announce_period: Duration,
+) {
+    // Handle request from NightshadeTask for confirmation on a payload.
+    // If the payload can't be built from the mempool task to fetch necessary data is spawned and the
+    // request is stored until it is ready.
+    let pool1 = pool.clone();
+    let control_tx1 = control_tx.clone();
+    let task = retrieve_payload_rx.for_each(move |(authority_id, hash)| {
+        info!(
+            "[{}] Payload confirmation for {} from {}",
+            pool1.authority_id.read().expect(""),
+            hash,
+            authority_id
+        );
+        if !pool1.contains_payload_snapshot(&hash) {
+            tokio::spawn(
+                payload_request_tx
+                    .clone()
+                    .send(PayloadRequest::BlockProposal(authority_id, hash))
+                    .map(|_| ())
+                    .map_err(|e| warn!(target: "pool", "Error sending message: {}", e)),
+            );
+
+            pool1.add_pending(authority_id, hash);
```

Comment:
`pending_snapshots` are not actually used, so this does nothing. Why do we need it?